File: backend/src/internal/models/event.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type Event struct {
	ID uint `json:"id" gorm:"primaryKey"`

	Type string `json:"type" gorm:"size:100"`

	Title       string  `json:"title" gorm:"size:500" validate:"required,min=2"`
	Category    *uint   `json:"category,omitempty"`
	Views       *uint   `json:"views,omitempty" gorm:"column:views"`
	Description string  `json:"description" gorm:"type:text"`
	OrderNum    uint    `json:"orderNum" gorm:"default:0"`
	MainImage   *string `json:"mainImage,omitempty" gorm:"type:text"`
	ImageUpload *string `json:"imageUpload,omitempty"`
	Level       uint    `json:"level" gorm:"default:0"`

	UserID uint `json:"userId"`
	User   User `json:"author" gorm:"constraint:OnUpdate:CASCADE,OnDelete:SET NULL;"`

	Status bool `json:"status"`

	DomainID *uint   `json:"domainId" gorm:"index"`
	Domain   *Domain `json:"domain" gorm:"constraint:OnUpdate:CASCADE,OnDelete:SET NULL;"`

	CreatedDate *time.Time `json:"createdDate,omitempty"`
	ShowFrom    time.Time  `json:"showFrom"`
	ShowTo      time.Time  `json:"showTo"`

	CreatedAt time.Time       `json:"createdAt"`
	UpdatedAt time.Time       `json:"updatedAt"`
	DeletedAt *gorm.DeletedAt `gorm:"index" json:"deletedAt,omitempty"`
}

File: frontend/src/src/actions/event.ts
import { gql } from "@apollo/client";

export const GET_TOP_EVENT = gql`
  query GetTopEvents {
    response: topEvents {
      id
      title
      description
      orderNum
      status
      showFrom
      showTo
      type
      domainId
      domain {
        id
        name
      }
      userId
      user {
        id
        name
        userid
      }
      level
      createdAt
      updatedAt
      deletedAt
    }
  }
`;
export const FILTER_EVENT = gql`
  query GetEvents(
    $filters: [Filter!]
    $orders: [Order!]
    $pagination: Pagination
  ) {
    response: getEvents(
      filters: $filters
      orders: $orders
      pagination: $pagination
    ) {
      events {
        id
        title
        description
        orderNum
        status
        showFrom
        showTo
        imageUpload
        category
        type
        views
        domainId
        userId  
        user {
          id
          name
          userid
        }
        createdDate
        level
        createdAt
        updatedAt
        deletedAt
      }
      total
    }
  }
`;

export const CREATE_EVENT = gql`
  mutation NewEvent($input: NewEventInput!) {
    response: createEvent(input: $input) {
      id
      title
      category
      views
      createdDate
      showFrom
      showTo
      description
      mainImage
      imageUpload
    }
  }
`;

export const UPDATE_EVENT = gql`
  mutation UpdateEvent($id: ID!, $input: UpdateEventInput!) {
    response: updateEvent(id: $id, input: $input) {
      id
      title
      showFrom
      showTo
      orderNum
      description
      status
      createdAt
      updatedAt
      deletedAt
    }
  }
`;

export const DELETE_EVENT = gql`
  mutation DeleteEvent($id: ID!) {
    response: deleteEvent(id: $id)
  }
`;
